src/tezos_interop/tezos_interop.ml

```diff
@@ -187,16 +204,45 @@ module Consensus = struct
     Listen_transactions.listen t ~rpc_node:t.rpc_node
       ~required_confirmations:t.required_confirmations
       ~destination:t.consensus_contract ~on_message
+  let fetch_discovery t validator_key_hash =
+    let {
+      rpc_node;
+      required_confirmations;
+      consensus_contract = _;
+      discovery_contract;
+      secret = _;
+      bridge_process = _;
+    } =
+      t in
+    let micheline_to_validators (michelson : (Michelson.t, string) result) =
+      let%ok michelson = michelson in
+      let michelson = Micheline.root michelson in
+      match michelson with
+      | Micheline.Prim (_, D_Pair, [_; String (_, uri)], _) ->
+        Ok (Uri.of_string uri)
+      | _ -> Error "Failed to parse storage micheline expression" in
+    let%await micheline_storage =
+      Fetch_big_map.run t ~required_confirmations ~rpc_node
+        ~contract_address:discovery_contract ~key:(Key_hash validator_key_hash)
+    in
+    Lwt.return (micheline_to_validators micheline_storage)
+
   let fetch_validators t =
     let {
       rpc_node;
       required_confirmations;
       consensus_contract;
+      discovery_contract = _;
       secret = _;
       bridge_process = _;
     } =
       t in
 
+    let rec traverse_result = function
+      | [] -> Ok []
+      | Ok x :: xs -> Result.bind (traverse_result xs) (fun xs -> Ok (x :: xs))
+      | Error e :: _ -> Error e in
+
```

Comment:
Woo monad transformers. 

What do we think about adding 
```
List.fold_result : ('a -> 'b -> ('a, 'c) result) -> 'a -> 'b list -> ('a, 'c) result list`
```
and 
```
List.fold_result_lwt : ('a -> 'b -> ('a, 'c) result Lwt.t) -> 'a -> 'b list -> ('a, 'c) result Lwt.t
```
To the `List` module, and then doing it in a single fold_result_lwt down below?

Reply:
@EduardoRFS done
@d4hines possibly useful, but since we already have `fold_left_ok` in there, and there's already `Lwt.all`, I'm not sure how useful it would be.